### PR TITLE
Component Group View with static header and without icons

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

Add an initial version of the visual component for displaying the Component Group View. The component contains a header (for displaying the Group Name) and a list of labels (for displaying the component names).

https://www.pivotaltracker.com/story/show/181724889

#### Visuals

A screenshot from a debug scene demonstrating the component:

Screenshot 2022-04-13 at 20 07 56

Screenshots from other debug scenes (`list_view` and `text_area`), demonstrating that the other components still display correctly:

Screenshot 2022-04-13 at 20 08 56

Screenshot 2022-04-13 at 20 08 48

A screenshot of the Node Searcher's list, showing that long entries in a ListView are now truncated, and an ellipsis character is added in place of removed characters:

Screenshot 2022-04-13 at 20 10 16

### Important Notes

 - Adding support for the text truncation feature in `ListView` required some changes in the`list_view::Entry`-related APIs.
 - An embedded font was added (DejaVuSans-Bold) for use in the Component Group View debug scene, and 5 unused embedded fonts were removed.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

**Comment by akavel** [comment]
Note: Moving this PR back to "Draft" state for a short while to extract the code for changing fonts via styles into a separate PR (suggested during the Demo).

**Comment by akavel** [comment]
~~Based on discussion with @farmaazon , we want first to add support for text values in style sheet, *then* the Component Group View. Thus, converting this PR back to draft again, for the time being.~~

After further discussion, the decision is reverted, we're starting with "ugly CGV", *then* will add text-value styles.

File: lib/rust/ensogl/component/Cargo.toml
Comment (farmaazon) [inline lib/rust/ensogl/component/Cargo.toml]: "Component Group View" is a name specific to Enso IDE. We should rename it to something more general, or make it a sub-crate of ide_view.

File: lib/rust/ensogl/component/component-group-view/src/lib.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/component-group-view/src/lib.rs]: These docs are a bit fuzzy. Perhaps due to fact, that the crate is in the wrong place; in context of "EnsoGL component" I don't know what are Enso components, and what are their groups.

File: lib/rust/ensogl/component/component-group-view/src/lib.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/component-group-view/src/lib.rs]: Please put last ) in a new line.

File: lib/rust/ensogl/component/component-group-view/src/lib.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/component-group-view/src/lib.rs]: I think it's better to write exact description in the main structure's docs, and refer to it from module's docs.

File: lib/rust/ensogl/component/component-group-view/src/lib.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/component-group-view/src/lib.rs]: Make a constant, put it on the top of the file.

Also I think the embedded-fonts crate should already contain constants with font names (so the typos would be detected during compilation). But if you think this refactoing would take a whole day or more, let's merge this PR first, and do the refactoring later.

File: lib/rust/ensogl/component/component-group-view/src/lib.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/component-group-view/src/lib.rs]: This structure is used in FRP, so I think it should be defined before (this is pattern followed in most files and personally I find it convenient).

File: lib/rust/ensogl/component/list-view/src/entry/list.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/list-view/src/entry/list.rs]: Mention the new parameter in the docs - it is not 100% clear of what the width is.

File: lib/rust/ensogl/component/list-view/src/entry.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/list-view/src/entry.rs]: Resize suggests that we set both width and height. Let it be `set_width`.

File: lib/rust/ensogl/component/text/src/component/area.rs
Comment (farmaazon) [inline lib/rust/ensogl/component/text/src/component/area.rs]: Add docs. The name does not clearly state what does it do.

Also, I'm not sure if it works well with editing text - not every operation does the full line redrawing, and that's the only place where we truncate the text.

I think we should not have "truncate" parameter, but just add endpoint "set_content_truncated" - this way we avoid some bad surprises.

File: lib/rust/ensogl/core/src/display/scene/layer.rs
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/scene/layer.rs]: At first glance, this comment does not look like explaining anything, rather like forgotten commented code. Please add some sentence here.

**Review by farmaazon — CHANGES_REQUESTED** [review]
1. There is a glitch near header background
Screenshot 2022-04-12 at 13 11 58
2. I mention somewhere in comments the problem with component-group-view inside EnsoGL - I think this component should be inside ide-view crate. Let put it in "app/gui/view/component-browser/component-group-view". (the "component-browser" can also be a crate, for now empty). The example should be in app/gui/view/debug-scene.
3. I know this is an "ugly" version, but could the colors be taken from design doc? Or, at least, could the header have a different color?

File: lib/rust/ensogl/component/text/src/component/area.rs
Comment (akavel) [reply lib/rust/ensogl/component/text/src/component/area.rs]: The truncation feature is not currently expected to work well with text editing in `text::Area`. Discussing this with @farmaazon, we considered three possible alternative solutions to the comment above:

 - a) Rename to `set_content_truncated(String, f32)` and add a comment explaining that this endpoint should not be used together with the text editing features.
 - b) Modify `text::Area` to require choosing a "mode", which would be one of three possible variants: editable, read-only, read-only-with-truncation.
 - c) Rename to `set_content_truncated(String, f32)`, and change the semantics of this operation from working in the presentation layer to operating on the data model. In other words, using `set_content_truncated` would be equivalent to truncating the string and then passing it to `set_content`. This would be also explained in a comment, and would require storing the full (un-truncated) text in other visual components (e.g. in `list_view::entry::Label` and the Component Group View).

Per @farmaazon's request, due to time constraints, option (c) will be chosen for now.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: We try not to repeat the names in the docs. The names such as "Enso" or "EnsoGL" do not have to appear here. It makes it easier in the future to refactor these names. Also:

1. "An EnsoGL implementation of a Component Group View." – this sounds like there was a non-EnsoGL implementation of "Component Group View". If so, where it is? Also, the first paragraph should explain what this file is about, but it does not. After reading it, I don't know what "Component Group View" really is.
2. "header (containing the name of a group of Enso components)", what else the header contains?
3. "[...] it (names of the components belonging to that group)", this sounds like the explanation is connected with "it", but it is not.
4. This docs should be linked to our design document.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: why is this needed?

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: Instead of too many imports, you should use `list_view::entry` in code.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: Why this thing is called "Component Group View" instead of just "Component Group"? I know that we have "List View", but this is because "List" is a data structure. Other components, like "Text Area" are not named "Text Area View". @farmaazon, what do you think of it?

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: if this is just a rect with not rounded corners, why dont you use plane instead? Plane is faster to render

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: what else you can set in header? IMO this should be called "set_header" instead.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: Please, try making the file readable from top-to-bottom. `HeaderGeometry` was not introduced yet. Move it above these definitions, if possible, please.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group-view/src/lib.rs]: this should not be hardcoded here.

File: app/gui/view/component-browser/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/src/lib.rs]: why is this needed?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: why is this needed?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: where do you use this extension?

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: use the new entry point macro - see EnsoGL demo scenes to learn more.

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: Please, make the demo scenes look not like created by a "developer" ;) Use some more subtle colors rather than pure green background.

File: build/run.js
Comment (wdanilo) [inline build/run.js]: How did it happen? I'm concerned right now. Is this number different on your system than other systems?

File: lib/rust/ensogl/component/list-view/src/entry/list.rs
Comment (wdanilo) [reply lib/rust/ensogl/component/list-view/src/entry/list.rs]: this parameter should also be named better. Is this `max_width_px`?

File: lib/rust/ensogl/component/list-view/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/component/list-view/src/lib.rs]: which macro requires it?

File: lib/rust/ensogl/component/list-view/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/component/list-view/src/lib.rs]: please format it the same was as rustfmt formats things

File: lib/rust/ensogl/component/text/src/component/area.rs
Comment (wdanilo) [reply lib/rust/ensogl/component/text/src/component/area.rs]: > The truncation feature is not currently expected to work well with text editing in text::Area 

Oh wow, then remove it. @farmaazon I'm really glad you've caught it! @akavel, if you are adding something, it has to work. Wither this has to work with editing the text, or this should not be integrated with text area. If we do not want to make it working with editing the text, the option c) is the only viable option to me right now.

**Review by wdanilo — CHANGES_REQUESTED** [review]

**Review by farmaazon — APPROVED** [review]
1. The glitch is currently fixed.
2. I leave an approval, so you will not be blocked in my absence. Please resolve all comments of @wdanilo and mine before merging.

File: lib/rust/ensogl/component/list-view/src/lib.rs
Comment (akavel) [reply lib/rust/ensogl/component/list-view/src/lib.rs]: Huh; while I was developing this PR, the compilation started failing at some point due to the recursion limit being exceeded, and Adam suggested that I increase the limit; but at the current state of the PR, this increase appears not needed anymore. I reverted it back to the original value.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (akavel) [reply app/gui/view/component-browser/component-group-view/src/lib.rs]: For some reason I don't understand, changing this line to use Plane instead:
```rust
            let shape = Plane().fill(color);
```
results in the component rendering with wrong colors - specifically, the color behind the ListView is correct, but the one in the header area is not:

Screenshot 2022-04-13 at 20 47 29

Would you have some idea or hint as to what could be the reason for this? I checked what will happen if I use `Plane` in the `shape_system` debug scene, by changing the `shape()` function as below:
```rust
pub fn shape() -> AnyShape {
    let shape = Plane().fill(color::Rgba(0.927, 0.937, 0.913, 1.0));
    shape.into()
}
```
and I also got a rectangle of a different color than expected:
Screenshot 2022-04-13 at 21 04 53

Whereas if I restore the `shape()` function to its original content, then change only the color, as below:
```rust
pub fn shape() -> AnyShape {
    let circle1 = Circle(50.px());
    let circle_bg = circle1.translate_x(-(50.0.px()));
    let circle_sub = circle1.translate_y(-(50.0.px()));
    let rect = Rect((100.0.px(), 100.0.px()));
    let shape = circle_bg + rect - circle_sub;
    let shape = shape.fill(color::Rgba::new(0.927, 0.937, 0.913, 1.0));
    shape.into()
}
```
then I can see the expected color:
Screenshot 2022-04-13 at 21 10 45

File: build/run.js
Comment (akavel) [reply build/run.js]: As discussed with @farmaazon, I'm lowering the limit as a result of deleting several unused embedded fonts.

File: app/gui/view/component-browser/component-group-view/src/lib.rs
Comment (wdanilo) [reply app/gui/view/component-browser/component-group-view/src/lib.rs]: Interesting. Seems like a bug. I will investigate that. But let's use rectangle now + create comment that it will be better to use plane, but there is some colouring bug. May be related to: https://github.com/enso-org/enso/pull/3384

File: app/gui/view/component-browser/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/lib.rs]: what is `t`? It is really poorly named variable

File: app/gui/view/debug_scene/component-group/src/lib.rs
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: what needs this?

**Review by wdanilo — APPROVED** [review]
Mateusz, this is really nice code and nicely done PR!

**Comment by akavel** [comment]
(*Process note:* the code-review is now Approved, therefore the PR is waiting for an Acceptance/QA Review by @vitvakatu  on Task https://www.pivotaltracker.com/story/show/181724889 before merge.)

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/181724889">Component Group View with static header and without icons - Pivotal Tracker</a></strong></div></blockquote>

**Review by vitvakatu — APPROVED** [review]
Reviewed and tested manually, QA Acceptance set to pass. Automated tests are not possible here.

**Pull request merged**
